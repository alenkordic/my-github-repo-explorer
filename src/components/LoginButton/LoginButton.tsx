import React from "react";
import { MenuItem, Typography, Button } from "@mui/material";
import { useSearchParams } from "react-router-dom";

import { useAuthContext } from "./../../contexts/auth.context";
import { getTokens } from "./../../services/api";
import { client_id, redirect_uri } from "./../../constants/enviroments";

interface TokensProps {
  accessToken: string;
  refreshToken: string;
}

const LoginButton = () => {
  // @ts-ignore
  const { login, logout, isAuthenticated } = useAuthContext();

  const [searchParams] = useSearchParams();
  const code = searchParams.get("code");

  if (code) {
    getTokens(code).then((res: TokensProps) => {
      if (res) {
        login(res.accessToken, res.refreshToken);
      }
    });
  }

  const handleLogout = () => {
    logout();
  };

  return (
    <div data-test="loginButton" data-cy="authButtonGroup">
      {isAuthenticated && (
        <MenuItem onClick={handleLogout} data-cy="logoutButton">
          <Typography textAlign="center">Logout</Typography>
        </MenuItem>
      )}

      {!isAuthenticated && (
        <Button
          sx={{ my: 2, color: "white", display: "block" }}
          data-cy="loginButton"
        >
          <a
            href={`https://github.com/login/oauth/authorize?scope=user&client_id=${client_id}&redirect_uri=${redirect_uri}`}
            style={{ textDecoration: "none", color: "inherit" }}
          >
            <Typography textAlign="center">Login with GitHub</Typography>
          </a>
        </Button>
      )}
    </div>
  );
};

export default LoginButton;
