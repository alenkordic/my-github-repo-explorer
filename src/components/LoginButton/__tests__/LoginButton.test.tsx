import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

import LoginButton from "../LoginButton";
import { AuthContext } from "../../../contexts/auth.context";

function getBtn(isAuth: boolean) {
  const value = {
    isAuthenticated: isAuth,
    logout: jest.fn(),
    login: jest.fn()
  };
  return (
    <AuthContext.Provider value={value}>
      <BrowserRouter>
        <LoginButton />
      </BrowserRouter>
    </AuthContext.Provider>
  );
}

test("Login btn when user is not authenticated.", () => {
  render(getBtn(false));
  expect(screen.getByText("Login with GitHub")).toBeTruthy();
});

test("Logout btn when user is authenticated.", () => {
  render(getBtn(true));
  expect(screen.getByText("Logout")).toBeTruthy();
});

// Snapshots
it("Login btn render correctly", () => {
  const loginBtn = getBtn(false);
  const tree = renderer.create(loginBtn).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Logout btn render correctly", () => {
  const logoutBtn = getBtn(true);
  const tree = renderer.create(logoutBtn).toJSON();
  expect(tree).toMatchSnapshot();
});
