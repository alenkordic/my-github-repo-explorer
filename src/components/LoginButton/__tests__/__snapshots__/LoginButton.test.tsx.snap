// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login btn render correctly 1`] = `
<div
  data-cy="authButtonGroup"
  data-test="loginButton"
>
  <button
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1q39md6-MuiButtonBase-root-MuiButton-root"
    data-cy="loginButton"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <a
      href="https://github.com/login/oauth/authorize?scope=user&client_id=Iv1.a22e2f19afba852b&redirect_uri=http://localhost:3000/"
      style={
        Object {
          "color": "inherit",
          "textDecoration": "none",
        }
      }
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
      >
        Login with GitHub
      </p>
    </a>
  </button>
</div>
`;

exports[`Logout btn render correctly 1`] = `
<div
  data-cy="authButtonGroup"
  data-test="loginButton"
>
  <li
    className="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-cy="logoutButton"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="menuitem"
    tabIndex={-1}
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
    >
      Logout
    </p>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>
</div>
`;
